Scrape multiple terms

ondenman opened the pull request:
This PR adds multiple terms to the scraper.

Row class has been amended to handle differing number of
columns by indexing the table columns from the right.

For example, the party name is always the last field. However its index
from the start varies between term 2016 and terms 2008 and 2012.

ondenman committed:
Rename Khurai to Page

Renamed to reflect that the class represents a page on wikipedia.

ondenman committed:
Move Table class to separate file

ondenman committed:
Move Row class to separate file

ondenman committed:
Move Page class to separate file

ondenman committed:
Add ignore file

ondenman committed:
Reference tds by negative index

The source table on the 2016 contains one more column than
the 2012 and 2008 term pages. As a result, indexing from
the left is not consistent However, indexing from the
right is (for these three terms).

For example, the party name is always the last field. However
its index from the start varies between term 2016 and terms
2008 and 2012.

https://en.wikipedia.org/wiki/List_of_MPs_elected_in_the_Mongolian_legislative_election,_2016
https://en.wikipedia.org/wiki/List_of_MPs_elected_in_the_Mongolian_legislative_election,_2012
https://en.wikipedia.org/wiki/List_of_MPs_elected_in_the_Mongolian_legislative_election,_2008

ondenman committed:
Scrape multiple terms

Adds 2012 and 2008 terms.

tmtmtmtm commented on lib/page.rb:
```diff
@@ -0,0 +1,23 @@
+
+  def table
+    page.xpath('.//h2/span[text()[contains(.,"Constituency")]]/following::table[1]')
+  end
```
This seems remarkably specific for a generic `Page` class. I suspect this should really live in a subclass.

tmtmtmtm commented on lib/row.rb:
```diff
@@ -0,0 +1,41 @@
+      term: term,
+      wikiname: wikiname
+    }
+  end
```
Is there a reason why you can't use the field-serializer for this?

tmtmtmtm commented on lib/table.rb:
```diff
@@ -0,0 +1,20 @@
+require 'nokogiri'
+
+class Table
```
Both this and Row are quite tightly coupled to the underlying implementation details. It would be a lot better to name the classes after what this represent, not how the data happens to currently be presented.

ondenman committed:
Rename Page#page to Page#noko

Giving method less specific name as a step
towards turning Page into a base class.

ondenman committed:
Add FieldSerializer to Page

ondenman committed:
Add TermPage class

Moves #members and #table from Class to subclass TermPage.

ondenman committed:
Use field serializer in TermPage

Replaces #members with :members field.

ondenman committed:
Update Ruby version to 2.3.1

For Array#to_h support as required by FieldSerializer.

ondenman committed:
Use FieldSerializer in Row

ondenman committed:
Move fields from Row to subclass MemberRow

Moves specificity to MemberRow.

ondenman committed:
Remove unnecessary requires

Scraper only needs to know about TermPage. Classes should
require dependencies themselves.

ondenman committed:
Rename Table to Members

New name is felt to be a better reflection of Class's purpose.

ondenman committed:
Rename Row to Member

Name felt to be a better representation of the Class's purpose.
Hides implementation.

ondenman committed:
Rename MemberRow to KhuralMember

New name is felt to be a better reflection of the class's
purpose: to represent a member of the State Great Khural of
Mongolia.

Hides implementation.

tmtmtmtm commented on lib/member.rb:
```diff
@@ -0,0 +1,14 @@
+require 'nokogiri'
```
Why does this need `nokogiri`? It doesn't seem to use it.

At a more fundamental level what _is_ this class? It seems to be trying to be a generic 'Member' class, but it's not clear what functionality it's providing, and it's quite specific about needing to be instantiated with some `tds`. It's certainly possible that there's a useful abstraction here to capture, but if there is it would be good to give it a name that reflects better what it is.

tmtmtmtm commented on lib/members.rb:
```diff
@@ -0,0 +1,21 @@
+require 'nokogiri'
```
Again, nothing here seems to actually need this.

tmtmtmtm commented on scraper.rb:
```diff
@@ -4,93 +4,24 @@
 require 'scraperwiki'
 require 'nokogiri'
```
I don't think you need nokogiri here either

tmtmtmtm commented on Gemfile:
```diff
@@ -14,5 +14,4 @@ gem "nokogiri"
 gem 'wikidata-client', '~> 0.0.7', require: 'wikidata'
-
-
+gem 'field_serializer', github: 'everypolitician/field_serializer'
```
You should using secure github links (per https://github.com/everypolitician/everypolitician/issues/518): 

```
git_source(:github) { |repo_name| "https://github.com/#{repo_name}.git" }
```

This will avoid the warning that `bundle install` is giving:

```
The git source `git://github.com/everypolitician/field_serializer.git` uses the `git` protocol, which transmits data without encryption. Disable this warning with `bundle config git.allow_insecure true`, or switch to the `https` protocol to keep your data secure.
```

tmtmtmtm commented on lib/khural_member.rb:
```diff
@@ -0,0 +1,23 @@
+  end
+
+  field :term do
+    '2016'
```
Surely this shouldn't be hard-coded to 2016, if it's being used for multiple terms?

tmtmtmtm commented on lib/khural_member.rb:
```diff
@@ -0,0 +1,23 @@
+  end
+
+  field :wikiname do
+    tds[1].xpath('.//a[not(@class="new")]/@title').text.strip
```
Shouldn't this be the same cell as the `name` — i.e. -4 — so that it works with both layouts?

tmtmtmtm reviewed (changes requested):
As well as the specific issues mentioned here, I'm finding the various levels of indirection here a little hard to follow, especially as various parts of it almost, but not quite, follow the pattern (e.g. `to_a` vs `to_h`). 

I'm about to push a reworking that I did as I was reviewing this. Let's discuss on Monday.

tmtmtmtm commented:
superseded by https://github.com/everypolitician-scrapers/mongolia-khurai-wp-multiple-terms/pull/3
